src/main/java/io/pravega/connectors/flink/source/reader/PravegaSourceReader.java

```diff
@@ -0,0 +1,125 @@
+
+/**
+ * A Pravega implementation of {@link SourceReader}.
+ *
+ * @param <T> The final element type to emit.
+ */
+public class PravegaSourceReader<T> extends SourceReaderBase<EventRead<ByteBuffer>, T, PravegaSplit, PravegaSplit>
+        implements ExternallyInducedSourceReader<T, PravegaSplit> {
+    private static final Logger LOG = LoggerFactory.getLogger(PravegaSourceReader.class);
+
+    private Optional<Long> checkpointId;
+
+    private PravegaSourceReader(
+            FutureCompletingBlockingQueue<RecordsWithSplitIds<EventRead<ByteBuffer>>> elementsQueue,
+            Supplier<PravegaSplitReader> splitReaderSupplier,
+            RecordEmitter<EventRead<ByteBuffer>, T, PravegaSplit> recordEmitter,
+            Configuration config,
+            SourceReaderContext context) {
+        super(elementsQueue, new PravegaFetcherManager<>(elementsQueue, splitReaderSupplier::get), recordEmitter, config, context);
+        checkpointId = Optional.empty();
+    }
+
+    /**
+     * Creates a new Pravega Source Reader instance.
+     * The PravegaSourceReader has a default recommended Flink implementation {@link SourceReaderBase}.
+     * It constructs with three major components, {@link PravegaSplitReader}, {@link PravegaFetcherManager}
+     * and {@link PravegaRecordEmitter}.
+     * Each reader will have a single threaded fetcher which will supply the split reader and assign all the splits
+     * assigned by the enumerator to it.
+     *
+     * @param splitReaderSupplier   The Pravega split reader supplier.
+     * @param recordEmitter         The Pravega Source reader record emitter.
+     * @param config                The Flink configuration.
+     * @param context               The Pravega Source reader context.
+     */
+    public PravegaSourceReader(
+            Supplier<PravegaSplitReader> splitReaderSupplier,
+            RecordEmitter<EventRead<ByteBuffer>, T, PravegaSplit> recordEmitter,
+            Configuration config,
+            SourceReaderContext context) {
+        this(new FutureCompletingBlockingQueue<>(config.getInteger(SourceReaderOptions.ELEMENT_QUEUE_CAPACITY)),
+                splitReaderSupplier,
+                recordEmitter,
+                config,
+                context);
+    }
+
+    @Override
+    public InputStatus pollNext(ReaderOutput output) throws Exception {
+        InputStatus inputStatus = super.pollNext(output);
+        checkpointId = ((PravegaRecordEmitter) recordEmitter).getAndResetCheckpointId();
+
+        // When the record is a checkpoint record, set the inputStatus to NOTHING_AVAILABLE
+        if (checkpointId.isPresent()) {
+            inputStatus = InputStatus.NOTHING_AVAILABLE;
+            LOG.trace("Source reader status: {}", inputStatus);
+        }
+        return inputStatus;
+    }
+
+    @Override
+    public List<PravegaSplit> snapshotState(long checkpointId) {
+        // Pravega doesn't support partial failover recover so just return null here for Source Reader
```

Comment:
Same question about partial failure, could we make the faulty reader offline for partial recovery?

Reply:
This also was my initial idea and I also tried but failed. It takes me some time and also discussions to Pravega client team to understand the gap here. I'll try to explain here.

For partial failure recovery, it would be handled by Flink that recover part of the DAG that connected to the faulty reader on the lineage and recover this region to the last checkpoint(more detail on [FLIP-1](https://cwiki.apache.org/confluence/display/FLINK/FLIP-1+%3A+Fine+Grained+Recovery+from+Task+Failures)). This means that a single reader needs to have the ability to rewind back to the last checkpoint position.

For Pravega API, it is not working like that. When the client calls the `readerOffline`, the reader group state manager would reassign the segments owned by the reader, and the other online readers would take over and continue reading. It behaves too smart. If we do the `readerOffline` with the FLIP-1 recovery, we would find that the downstream operators in Flink would have the state stored in Flink checkpoint, while the source has already go beyond the Pravega checkpoint, which leads to duplicates.

Reply:
I see, you are saying that partial recovery expects recovery of the faulty reader _r_ to roll back to a specific point, which is the last checkpoint position for the segments previously owned by _r_. if that's the case, then could we pass a  `Position` object as part of the `readerOffline` call that corresponds to that position? I guess the follow up question is how we build such a `Position` object. I wonder if it is possible to derive a position from the checkpoint object.

Reply:
Not just that, after the reader offline, the other readers will take over the segment and continue reading, this would cause duplicates even if we can roll back the failed reader to the position. Therefore, it is not a valid partial recovery.

Reply:
I'll approve to unblock the work, but we should have an issue to discuss the possibility of enabling partial recovery. I'm concerned that it will be perceived as a disadvantage of the model that Pravega is following.

Reply:
I have created #663 to track this issue, but it is true that it is now a disadvantage of the model